visualization_msgs/msg/Marker.msg

```diff
@@ -56,6 +56,16 @@ geometry_msgs/Point[] points
 # NOTE: alpha is not yet used
 std_msgs/ColorRGBA[] colors
 
+# Only used if the marker has a UV mapping defined for it or one is defined using uv_coordinates.
+# A URI to where the texture is accessible.
+string texture_map
```

Comment:
Hmmm I agree but I don't know if it's a good idea to store the image in the message due to how large the image may be (especially if sent with every message). Another strategy could be to have the textures defined separately via another message stream and then the marker system can just apply those textures via a unique identifier present in both messages? This would also allow the removal of `update_texture` as that will not be a constant stream (only sends messages when changes are to be made or a new texture should be loaded) :thinking:.

Reply:
+1 for having the texture data in a message somewhere that can be recorded.

If we're worried about the marker message being too large because of the texture, then I think a separate texture stream makes sense (either a message or service used to associated a texture with some ID).

Reply:
I started working down the route of texture via an alternate stream. This ended up being a lot more complicated than anticipated for a number of reasons:

1. If we want RViz to apply a texture by topic (a marker specifies a topic which contains the texture to be used), the number of topics will grow quickly which doesn't scale well.
2. If we want to define a static `/texture` topic, we have to figure out a way to guarantee that RViz receives all published messages per `texture_id` before it is started.
3. There are race conditions to applying the textures though this can be mitigated as Ogre has the ability to check for a texture identifier (so can just wait until it has been loaded and ignore otherwise).

My preference is to stick with the URI method. We already have a history of using URIs to load visual messages without the ability to record (meshes). When the resource is unavailable, the mesh handler just throws a warning and does not show anything. In the case of markers and unloaded textures, the marker will still render based on the defined color values just without an applied texture. In addition, I can add the warning regarding the inability to load the given URI. I see this as a fine fallback behavior and it can be stated clearly in the message definition.

At a later date, we can add support for record-able messages via a more complex URI handler that parses a `ros_texture_topic://` prefix or something of the like.

Reply:
I noticed that `mesh_tools` implemented the [texture loading as a service](https://github.com/uos/mesh_tools/blob/e6f9a8cb69644c756a262f626ffe32118437b38c/rviz_map_plugin/src/MeshDisplay.cpp#L901) which is definitely doable just a matter of if the added complexity is the correct approach.

Reply:
> My preference is to stick with the URI method. We already have a history of using URIs to load visual messages without the ability to record (meshes). When the resource is unavailable, the mesh handler just throws a warning and does not show anything. In the case of markers and unloaded textures, the marker will still render based on the defined color values just without an applied texture. In addition, I can add the warning regarding the inability to load the given URI. I see this as a fine fallback behavior and it can be stated clearly in the message definition.

Just because we've taken the shortcut in the past doesn't mean it's good to do again. This is something that I remember being a common question on ROS Answer. (Why are my meshes all white?) And I think that the only things that we currently use meshes on are generally urdfs which also have URI style lookups and the workaround that we used with them is to do things like publish them on latched topics.

The service at least is reasonable for providing the same abstraction level. However, I think I need to understand the use cases we're trying to cover here. Are we talking about completely static textures that are reused over and over again. Or are we talking about dynamic textures that are generated based on runtime content. 

Note that we had a previous instance of this for tranforms and a transform server and we optimized to use numeric frame ids instead of strings. And it's much more efficient, but requiring the lookup service to resolve frame names completely compromised the experience. 

Coming into this I think that the basic consideration is that everything should be contained in the message. And it's only if that causes larger overhead that we should then consider increasing the complexity of the system via remote calls or caching. 

I see that there's potential use cases ranging from drawing a TV with live video feed on the surface. To drawing blades of grass with a realistic green pattern. What will be the overhead cost of resending the data in each message versus providing parallel infrastructure to send, store, and query for the data?

Reply:
> This is something that I remember being a common question on ROS Answer. (Why are my meshes all white?)

But this has to do with URDF, which I think is a bit different because the meshes are also specified by URI (not embedded) so the issue would be "I don't see any meshes". If their meshes are white, that means the textures the meshes reference cannot be located, which is a different issue and wouldn't be solved by embedding the mesh data somewhere in a message. I think because of this you're always going to have issues with locating resources for playing back old data, and using `file://` URIs isn't the only problem.

I've encountered the white meshes thing when tf wasn't working, which for some reason would result in a bundle of meshes at the origin, all white. So there might be other reasons for this error case.

> And I think that the only things that we currently use meshes on are generally urdfs which also have URI style lookups and the workaround that we used with them is to do things like publish them on latched topics.

When do we publish meshes used by URDF on topics? Is that common practice, because I've never seen it. Can a URDF even reference a topic?

Also, Marker messages already have URI for meshes. When @gbalke asked me about this I suggested sticking with the URI approach since it would be consistent with that, see:

>  The format is the URI-form used by resource_retriever, including the package:// syntax.

-- https://wiki.ros.org/rviz/DisplayTypes/Marker#Mesh_Resource_.28MESH_RESOURCE.3D10.29_.5B1.1.2B-.5D

Plus I think that if you support URIs then later you could add support for textures and/or meshes over topics or services.

The topics are complicated to implement and ensure textures are received by rviz (durability transient-local isn't a ready solution imo) and the services based solution is not record-able anyway (at this time).

Reply:
@tfoote I don't think we can actually decide for the user what the best approach is in this scenario. There's a reason to send the raw image in the message (for minimum latency/synchronization in a video streaming example) and there's reason to have it in a request format (very large textures for high fidelity environments). I think the best option is to provide a method to do all of the above. I think the URI method is still the best method due to the ability to extend to meet needs down the road.

In the case of wanting distributed resources, the user can simply specify a URL or file per the default URI behavior. We may extend this behavior to support two additional modes:
`raw_image_(format)://` followed by an encoded image.
`texture_<topic or id>://` followed by the topic or unique ID to be loaded (implementation similar to mesh_tools).

With this, we can state the performance and logging (rosbag) trade-offs that come along with each method. The benefit of this is the flexibility for the user to decide as I don't think we'll be able to "one-size-fits-all" this scenario.

Reply:
I would be happy to add `sensor_msgs/image` if not for the issue of sending the texture every time adding a significant amount of overhead. I also don't think we can avoid this else we have a race condition where the visualizer has to be started (and listening) before the first message is sent else miss the message containing the image information.